src/psyclone/psyir/nodes/structure_reference.py

```diff
@@ -233,13 +232,79 @@ def get_signature_and_indices(self):
 
         '''
         # Get the signature of self:
-        my_sig, my_index = \
-            super(StructureReference, self).get_signature_and_indices()
+        my_sig, my_index = super().get_signature_and_indices()
         # Then the sub-signature of the member, and indices used:
         sub_sig, indices = self.children[0].get_signature_and_indices()
         # Combine signature and indices
         return (Signature(my_sig, sub_sig), my_index + indices)
 
+    @property
+    def datatype(self):
+        '''
+        Walks down the list of members making up this reference to determine
+        the type that it refers to.
+
+        In order to minimise code duplication, this method also supports
+        ArrayOfStructuresReference by simply allowing for the case where
+        the starting reference is to an Array.
+
+        :returns: the datatype of this reference.
+        :rtype: :py:class:`psyclone.psyir.symbols.DataType`
+
+        '''
+        dtype = self.symbol.datatype
+
+        if isinstance(dtype, ArrayType):
+            dtype = dtype.intrinsic
+
+        if isinstance(dtype, DataTypeSymbol):
+            dtype = dtype.datatype
+
+        if isinstance(dtype, (DeferredType, UnknownType)):
+            # We don't know the type of the symbol that defines the type
+            # of this structure.
+            return dtype
+
+        # We do have the definition of this structure - walk down it.
+        cursor = self
+        cursor_type = dtype
+        shape = []
+        # The next two lines are required for when this method is called
+        # for an ArrayOfStructuresReference.
+        if isinstance(cursor, ArrayMixin) and cursor.shape:
+            shape = cursor.shape
+
+        try:
+            while cursor.member:
+                cursor = cursor.member
+                cursor_type = cursor_type.components[cursor.name].datatype
+                if isinstance(cursor_type, (UnknownType, DeferredType)):
+                    return DeferredType()
```

Comment:
This is a tricky one. I went for DeferredType because at this point we're working our way down a structure access. Unless this UnknownType is at the bottom then it *won't* be the type of the access.

Reply:
Since this is a towards PR I will let this pass to see it in action, but my worry with deferredtype is that it could imply that the datatype could be resolved if the proper datatype information is brought into scope from the modules, but if it comes from an UnknownType this isn't true.